test_data/std/btree_test.asm

```diff
@@ -0,0 +1,92 @@
+use std::btree::new;
+use std::btree::insert;
+use std::btree::search;
+use std::btree::CmpResult;
+use std::btree::BTree;
+use std::utils::Option;
+
+let cmp: int, int -> CmpResult = |a, b|
+    if a < b {
+        CmpResult::Less
+    } else {
+        if a == b {
+            CmpResult::Equal
+        } else {
+            CmpResult::Greater
+        }
+    };
+let b1 = new::<int, string>();
+let b2 = insert(b1, (1, "a"), cmp);
+let b3 = insert(b2, (8, "b"), cmp);
+let b4 = insert(b3, (4, "c"), cmp);
+let b5 = insert(b4, (2, "d"), cmp);
+let b6 = insert(b5, (9, "e"), cmp);
+let b7 = insert(b6, (3, "f"), cmp);
+let b8 = insert(b7, (7, "g"), cmp);
+let b9 = insert(b8, (5, "h"), cmp);
+let b10 = insert(b9, (6, "i"), cmp);
+let b11 = insert(b10, (0, "j"), cmp);
+let b12 = insert(b11, (10, "k"), cmp);
+// this one replaces
+let b13 = insert(b12, (4, "l"), cmp);
+let b14 = insert(b13, (6, "m"), cmp);
+
+let one: int = 1;
+let false: bool = one == 0;
+
+use std::debug::print;
+use std::debug::println;
+let spaces: int -> string = |n|
+    if n == 0 {
+        ""
+    } else {
+        spaces(n - 1) + " "
+    };
```

Comment:
Hm, I don't think this would be more efficient. We don't cache intermediate expressions (also because of side-effects).

But what would be more efficient is the following:
```
let spaces: int -> string = |n| if n == 0 { "" } else {
  let s = spaces(k / 2);
  match n % 2 {
    0 => s + s,
    1 => s + s +  " ",
  }
};
```